src/io/video.js

```diff
@@ -0,0 +1,332 @@
+         * @type {number}
+         */
+        this._skinId = -1;
+
+        /**
+         * The Scratch Renderer Skin object.
+         * @type {Skin}
+         */
+        this._skin = null;
+
+        /**
+         * Id for a drawable using the video's skin that will render as a video
+         * preview.
+         * @type {Drawable}
+         */
+        this._drawable = -1;
+    }
+
+    static get FORMAT_IMAGE_DATA () {
+        return 'image-data';
+    }
+
+    static get FORMAT_CANVAS () {
+        return 'canvas';
+    }
+
+    /**
+     * Dimensions the video stream is analyzed at after its rendered to the
+     * sample canvas.
+     * @type {Array.<number>}
+     */
+    static get DIMENSIONS () {
+        return [480, 360];
+    }
+
+    /**
+     * Order preview drawable is inserted at in the renderer.
+     * @type {number}
+     */
+    static get ORDER () {
+        return 1;
+    }
+
+    /**
+     * Request video be enabled.  Sets up video, creates video skin and enables preview.
+     *
+     * ioDevices.video.requestVideo()
+     *   .then(({ release }) => {
+     *     this.releaseVideo = release;
+     *   })
+     *
+     * @return {Promise.<VideoRequest>} A request object with a "release" property that
+     *      should be called when you are done with the video.
+     */
+    requestVideo () {
+        const io = this;
+        const request = {
+            release () {
+                const index = io._requests.indexOf(request);
+                if (index > -1) {
+                    io._requests.splice(index, 1);
+                }
+                if (io._requests.length === 0) {
+                    io._disablePreview();
+                    // by clearing refs to video and track, we should lose our hold over the camera
+                    io._video = null;
+                    io._track = null;
+                }
+            }
+        };
+
+        if (this.videoReady) {
+            this._requests.push(request);
+            return Promise.resolve(request);
+        }
+
+        if (this._lastSetup) {
+            return this._lastSetup.then(() => {
+                this._requests.push(request);
+                return request;
+            });
+        }
+
+        this._lastSetup = this._setupVideo()
+            .then(() => {
+                this._setupPreview();
+                this._requests.push(request);
+                this._lastSetup = null;
+                return request;
+            }, err => {
+                this._lastSetup = null;
+                throw err;
+            });
+        return this._lastSetup;
+    }
+
+    /**
+     * Create a video stream.
+     * Should probably be moved to -render or somewhere similar later
+     * @private
+     * @return {Promise} When video has been received, rejected if video is not received
+     */
+    _setupVideo () {
+        this._video = document.createElement('video');
+        return new Promise((resolve, reject) => {
```

Comment:
Handle the promise wrapped singleton storage in here. `if (!this._singleVideo) {this._singleVideo = ...;} return this._singleVideo;`.